File: src/java/test/src/androidTest/java/com/tencent/wcdbtest/winq/StatementDeleteTest.java
// Created by chenqiuwen on 2023/4/11.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdbtest.winq;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import com.tencent.wcdb.winq.*;
import static com.tencent.wcdbtest.base.WinqTool.winqEqual;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class StatementDeleteTest {
    @Test
    public void test() {
        Column column1 = new Column("column1");
        Column column2 = new Column("column2");

        winqEqual(genStatementDelete().where(column1.gt(1)),
                "DELETE FROM testTable WHERE column1 > 1");

        winqEqual(new StatementDelete().deleteFrom(new QualifiedTable("testTable").of("testSchema"))
                .where(column1.gt(1)),
                "DELETE FROM testSchema.testTable WHERE column1 > 1");

        winqEqual(genStatementDelete().orderBy(
                new OrderingTerm[]{column1.order(Order.Asc), column2.order(Order.Desc)}),
                "DELETE FROM testTable ORDER BY column1 ASC, column2 DESC");

        winqEqual(genStatementDelete().limit(1), "DELETE FROM testTable LIMIT 1");

        winqEqual(genStatementDelete().limit(1, 2), "DELETE FROM testTable LIMIT 1, 2");

        winqEqual(genStatementDelete().limit(1).offset(3), "DELETE FROM testTable LIMIT 1 OFFSET 3");

        winqEqual(genStatementDelete().where(column1.gt(1))
                .orderBy(new OrderingTerm[]{column1.order(Order.Asc), column2.order(Order.Desc)})
                        .limit(1).offset(2),
                "DELETE FROM testTable WHERE column1 > 1 ORDER BY column1 ASC, column2 DESC LIMIT 1 OFFSET 2");

        CommonTableExpression cte = new CommonTableExpression("tempTable")
                .as(new StatementSelect().select("testColumn").from("table1"));
        winqEqual(new StatementDelete().with(cte).deleteFrom("table2"),
                "WITH tempTable AS(SELECT testColumn FROM table1) DELETE FROM table2");
        winqEqual(new StatementDelete().withRecursive(cte).deleteFrom("table2"),
                "WITH RECURSIVE tempTable AS(SELECT testColumn FROM table1) DELETE FROM table2");
    }

    private StatementDelete genStatementDelete() {
        return new StatementDelete().deleteFrom("testTable");
    }
}

File: src/rust/examples/tests/winq/statement_delete_test.rs
#[cfg(test)]
pub mod statement_delete_test {
    use crate::base::winq_tool::WinqTool;
    use wcdb::winq::column::{Column, ColumnTrait};
    use wcdb::winq::expression_operable::ExpressionOperableTrait;
    use wcdb::winq::ordering_term::Order;
    use wcdb::winq::statement_delete::StatementDelete;

    #[test]
    pub fn test() {
        let test_table = "testTable";
        let statement = StatementDelete::new();

        let test = statement.delete_from(test_table);
        WinqTool::winq_equal(test, "DELETE FROM testTable");

        let column1 = Column::new("column1", None);
        let test = statement.where_(&column1.gt(100));
        WinqTool::winq_equal(test, "DELETE FROM testTable WHERE column1 > 100");

        let test = statement.limit(100);
        WinqTool::winq_equal(test, "DELETE FROM testTable WHERE column1 > 100 LIMIT 100");

        let column2 = Column::new("column2", None);
        let order1 = column1.order(Order::Asc);
        let order2 = column2.order(Order::Desc);
        let order_vec = vec![&order1, &order2];
        let test = statement.order_by(order_vec);
        WinqTool::winq_equal(test, "DELETE FROM testTable WHERE column1 > 100 ORDER BY column1 ASC, column2 DESC LIMIT 100");

        let test = statement.offset(100);
        WinqTool::winq_equal(test, "DELETE FROM testTable WHERE column1 > 100 ORDER BY column1 ASC, column2 DESC LIMIT 100 OFFSET 100");
    }

    #[test]
    pub fn limit_from_to() {
        let test_table = "testTable";
        let statement = StatementDelete::new();

        let test = statement.delete_from(test_table).limit_range(100, 200);
        WinqTool::winq_equal(test, "DELETE FROM testTable LIMIT 100, 200");
    }
}
